contracts/marketpools.js

```diff
@@ -0,0 +1,348 @@
+      lp.quoteQuantity = api.BigNumber(lp.quoteQuantity).plus(quoteQuantity);
+      await api.db.update('liquidityPosition', lp);
+    } else {
+      const newlp = {
+        account: api.sender,
+        tokenPair,
+        baseQuantity,
+        quoteQuantity,
+      };
+      await api.db.insert('liquidityPosition', newlp);
+    }
+
+    // deposit requested tokens to contract
+    const baseRes = await api.executeSmartContract('tokens', 'transferToContract', { symbol: baseSymbol, quantity: baseQuantity, to: 'marketpools' });
+    const quoteRes = await api.executeSmartContract('tokens', 'transferToContract', { symbol: quoteSymbol, quantity: quoteQuantity, to: 'marketpools' });
+    if (!api.assert(baseRes.errors === undefined && quoteRes.errors === undefined, 'deposit transfer errors')) return;
+    updatePoolStats(pool, baseQuantity, quoteQuantity, false);
+    api.emit('addLiquidity', { memo: `Add ${baseSymbol} and ${quoteSymbol}` });
+  }
+};
+
+actions.removeLiquidity = async (payload) => {
+  const {
+    tokenPair,
+    baseQuantity,
+    quoteQuantity,
+    isSignedWithActiveKey,
+  } = payload;
+
+  if (!api.assert(isSignedWithActiveKey === true, 'you must use a transaction signed with your active key')
+    || !api.assert(baseQuantity && api.BigNumber(baseQuantity).gt(0), 'invalid baseQuantity')
+    || !api.assert(quoteQuantity && api.BigNumber(quoteQuantity).gt(0), 'invalid quoteQuantity')
+    || !await validateTokenPair(tokenPair)) {
+    return;
+  }
+
+  const [baseSymbol, quoteSymbol] = tokenPair.split(':');
+  const pool = await api.db.findOne('pools', { tokenPair });
+  if (api.assert(pool, 'no existing pool for tokenPair')) {
+    if (!api.assert(api.BigNumber(pool.baseQuantity).gt(0)
+      && api.BigNumber(pool.quoteQuantity).gt(0), 'insufficient liquidity')
+      || !validateLiquiditySwap(pool, baseQuantity, quoteQuantity)) return;
+
+    const lp = await api.db.findOne('liquidityPosition', { account: api.sender, tokenPair });
+    if (api.assert(lp, 'no existing liquidity position')
+      && api.assert(api.BigNumber(lp.baseQuantity).minus(baseQuantity).gte(0)
+      && api.BigNumber(lp.quoteQuantity).minus(quoteQuantity).gte(0), 'not enough liquidity in position')) {
+      lp.baseQuantity = api.BigNumber(lp.baseQuantity).minus(baseQuantity);
+      lp.quoteQuantity = api.BigNumber(lp.quoteQuantity).minus(quoteQuantity);
+      if (api.BigNumber(lp.baseQuantity).eq(0) && api.BigNumber(lp.quoteQuantity).eq(0)) {
+        await api.db.remove('liquidityPosition', lp);
+      } else {
+        await api.db.update('liquidityPosition', lp);
+      }
+
+      await api.transferTokens(api.sender, baseSymbol, baseQuantity, 'user');
+      await api.transferTokens(api.sender, quoteSymbol, quoteQuantity, 'user');
+      updatePoolStats(pool, -baseQuantity, -quoteQuantity, false);
+      api.emit('removeLiquidity', { memo: `Remove ${baseSymbol} and ${quoteSymbol}` });
```

Comment:
Perhaps the emission should also contain the quantities removed?

Reply:
I was thinking it would be redundant if those amounts will already be emitted in the transferToContract events.

Reply:
Good point, wasn't thinking about that.  Should be fine as-is then, no change needed.